File: Leetcode Daily Challenge/013-July-2024/002-intersection-of-two-arrays-ii.py
# https://leetcode.com/problems/intersection-of-two-arrays-ii/




class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        # return self.approach_1(nums1, nums2)

        return self.approach_2(nums1, nums2)

    # TC : O(N)
    # SC : O(N)
    def approach_1(self, nums1, nums2):
        freq = [0] * 1001
        
        for num in nums2:
            freq[num] += 1
        
        result = []
        for num in nums1:
            if freq[num] > 0:
                freq[num] -= 1
                result.append(num)

        return result

    # TC : O(max(NlogN, MlogM)
    # SC : O(1)
    def approach_2(self, nums1, nums2):
        nums1.sort()
        nums2.sort()
        result = []
        i, j, m, n = 0, 0, len(nums1), len(nums2)
        while i < m and j < n:
            if nums1[i] == nums2[j]:
                result.append(nums1[i])
                i += 1
                j += 1
            elif nums1[i] < nums2[j]:
                i += 1
            else:
                j += 1
        return result

File: Leetcode Daily Challenge/013-July-2024/002-intersection-of-two-arrays-ii.java
// https://leetcode.com/problems/intersection-of-two-arrays-ii/



import java.util.*;

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        // return approach_1(nums1, nums2);

        return approach_2(nums1, nums2);
    }

    // TC : O( max(N*log(N), M*log(M)) )
    // SC : O( 1 )
    private static int[] approach_2(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;

        Arrays.sort(nums1);
        Arrays.sort(nums2);

        int i = 0;
        int j = 0;

        List<Integer> result = new ArrayList<>();

        while (i < m && j < n) {
            if (nums1[i] == nums2[j]) {
                result.add(nums1[i]);
                i ++;
                j ++;

            } else if (nums1[i] < nums2[j]) {
                i ++;
            } else j ++;
        }

        int[] finalResult = new int[result.size()];
        for (int k = 0; k < result.size(); k ++) {
            finalResult[k] = result.get(k);
        }

        return finalResult;
    }

    // TC : O(N)
    // SC : O(1001)
    private static int[] approach_1(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;
        
        int[] freq = new int[1001];

        for (int num : nums2) freq[num] ++;

        List<Integer> result = new ArrayList<>();
        for (int num : nums1) {
            if (freq[num] > 0) {
                freq[num] --;
                result.add(num);
            }
        }

        int[] resultArray = new int[result.size()];  // convert the list into array
        for (int i = 0; i < result.size(); i ++) {
            resultArray[i] = result.get(i);
        }
        return resultArray;
    }
}